### PR TITLE
Command family smart contract

### PR DESCRIPTION
**Pull request by isabeltomb**
This PR updates the command family and adds a sabre compatible command smart contract.

The `make_command_transaction` method is moved behind a separate experimental feature to allow for the command smart contract to be compiled into wasm.

The command family transaction handler is updated to use namespaces and a method is added to generate the sha512 hash of "command".

The necessary information is added to the docker files to allow for the command smart contract scar file to be build and archived in Jenkins.

### **Testing:**

1. Checkout the branch in this PR which adds the CLI subcommands for interacting with the command family smart contract: https://github.com/hyperledger/transact/pull/126
2. Start two splinter nodes and create a circuit between them with scabbard services
3. Download the command family smart contract scar file found here: https://build.sawtooth.me/job/Transact-Hyperledger/job/transact/job/PR-127/1/artifact/build/scar/
4. Upload the command family smart contract to the circuit using the scabbard CLI
5. Create a namespace registry for the contract using the namespace "06abbc"
```
scabbard ns create 06abbc \
    --owners <public_key> \
    --key <path_to_private_key> \
    --url http://0.0.0.0:8080 \
    --service-id <circuit-id>::<scabbard-service-id>
```
6. Add read and write permissions for the contract
```
scabbard perm 06abbc command --read --write \
    --key <path_to_private_key> \
    --url http://0.0.0.0:8080 \
    --service-id <circuit-id>::<scabbard-service-id>
```
7. Start the transact CLI using the branch from step 1, if you are using a docker file for your splinter nodes you can add something similar to the following to your docker file: 
```
  transact-cli:
    image: transact-cli
    build:
      context: ../transact
      dockerfile: ./cli/Dockerfile
      args:
      - CARGO_ARGS=${CARGO_ARGS}
      - REPO_VERSION=0.4.1-dev
    container_name: transact-cli
    hostname: transact-cli
```
and then set the environment variable `CARGO_ARGS="-- --features=experimental"` and then run `docker-compose -f <docker-file-name>.yaml run transact-cli bash` in a new terminal window to start the transact CLI 
(alternatively the transact commands can be run from the transact/cli directory as `cargo run --features=experimental ...`)

8. Run a set-state command
```
transact command set-state \
  --key /alice.priv
  --target "http://0.0.0.0:8080/scabbard/<circuit-id>/<scabbard-service-id>"
  --state-entry 06abbcb16ed7d24b3ecbd4164dcdad374e08c0ab7518aa07f9d3683f34c2b3c67a1583:state-value
```
9. Check the splinterd logs to see if the transaction was successfully submitted
10. Run a show-state command
```
transact command show-state \
  --key /alice.priv
  --target "http://0.0.0.0:8080/scabbard/<circuit-id>/<scabbard-service-id>"
  --address 06abbcb16ed7d24b3ecbd4164dcdad374e08c0ab7518aa07f9d3683f34c2b3c67a1583
  --text
```
11. Check that the "state-value" is returned

### CODE REVIEW

File: examples/sabre_command/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+    let txn_pair = txn
+        .into_pair()
+        .map_err(|err| SabreApplyError::InvalidTransaction(err.to_string()))?;
+
+    let mut context = SabreContext { context };
+
+    let handler = CommandTransactionHandler::new();
+
+    // wrap SabreTransactionContext into SabreContext that can be passed to a transact
```
Comment (agunde406) [inline examples/sabre_command/src/main.rs:62]: This comment should either be removed or moved above where you make the SabreContext

File: Cargo.toml [hunk 2]
```diff
@@ -6,6 +6,6 @@ members = [
   "cli"
   ]
 
-# Exclude sabre_smallbank to avoid incompatible features `sabre-compat` and
-# `log`
-exclude = ["examples/sabre_smallbank"]
+# Exclude sabre_smallbank and sabre_command to avoid incompatible features
```
Comment (agunde406) [inline Cargo.toml:9]: Typo in commit message, Exclue -> Exclude

**Review by agunde406 — CHANGES_REQUESTED** [review]
Tested locally and looked good. Just a few comments.

File: libtransact/Cargo.toml [hunk 3]
```diff
@@ -102,6 +102,7 @@ experimental = [
     "database-redis",
     "database-sqlite",
     "family-command",
+    "family-command-cylinder",
```
[the remark this answers is not in the document]
Comment (isabeltomb) [reply libtransact/Cargo.toml]: Changed this to "family-command-workload"

File: libtransact/src/families/command.rs [hunk 4]
```diff
@@ -122,6 +128,7 @@ impl TransactionHandler for CommandTransactionHandler {
     }
 }
 
+#[cfg(feature = "family-command-cylinder")]
```
[the remark this answers is not in the document]
Comment (isabeltomb) [reply libtransact/src/families/command.rs]: Made a command/workload directory and moved `make_command_transaction` there

File: libtransact/src/families/command.rs [hunk 5]
```diff
@@ -39,14 +40,20 @@ const COMMAND_VERSION: &str = "0.1";
...
 pub struct CommandTransactionHandler {
     versions: Vec<String>,
+    namespaces: Vec<String>,
 }
 
 impl CommandTransactionHandler {
     pub fn new() -> Self {
         CommandTransactionHandler {
             versions: vec![COMMAND_VERSION.to_owned()],
+            namespaces: vec![get_command_prefix()],
```
[the remark this answers is not in the document]
Comment (isabeltomb) [reply libtransact/src/families/command.rs]: Fixed

File: libtransact/src/families/command.rs [hunk 6]
```diff
@@ -35,7 +35,7 @@ use crate::protocol::transaction::TransactionPair;
 use crate::protos::{FromBytes, IntoBytes};
 
 const COMMAND_FAMILY_NAME: &str = "command";
-const COMMAND_VERSION: &str = "0.1";
+const COMMAND_VERSION: &str = "1.0";
```
[the remark this answers is not in the document]
Comment (isabeltomb) [reply libtransact/src/families/command.rs]: Fixed

File: examples/sabre_command/src/main.rs (hunk 1 above)
Comment (isabeltomb) [reply examples/sabre_command/src/main.rs]: Moved to the correct spot

File: Cargo.toml (hunk 2 above)
Comment (isabeltomb) [reply Cargo.toml:9]: Fixed

**Review by agunde406 — APPROVED** [review]

**Review by shannynalayna — APPROVED** [review]

**Pull request merged**